### ISSUE TITLE
Rewrite Context Search with WebExtensions APIs by end of 2017

### ISSUE THREAD

**Issue by benbasson** [benbasson/contextsearch#27]
Mozilla have announced plans to phase out all non-WebExtension add-ons by the end of 2017. While I consider this decision to be utterly insane (and half expect them to end up extending that deadline, if they have any users left to appease at that point), it's safe to assume that this add-on will have to be rewritten or will cease to exist at the end of 2017 or shortly after.

Context menu APIs are present (although very primitive) and APIs for search plugins are seemingly absent at the moment. If and when these come along, a rewrite will be more viable.

Current blockers in the WebExtensions API:

- [Bug 1352598 - Add an API to list installed search engines](https://bugzilla.mozilla.org/show_bug.cgi?id=1352598)
- <del>[Bug 1302742 - Pass modifier (shift,cmd,...) to onclick callback for chrome.contextMenus.create()](https://bugzilla.mozilla.org/show_bug.cgi?id=1302742)</del>

Triaged and approved non-blockers that would make for a better user experience (likely to be worked on and implemented):

- [Bug 1320462 - WebExtensions: Add ability to set access key to context menu item](https://bugzilla.mozilla.org/show_bug.cgi?id=1320462)
- [Bug 1215376 - Implement onBeforeShow event](https://bugzilla.mozilla.org/show_bug.cgi?id=1215376)
- [Bug 1268401 - Add APIs to manage search engines for the web search bar/location bar](https://bugzilla.mozilla.org/show_bug.cgi?id=1268401)

Untriaged non-blockers that would make for a better user experience:

- <del>[Bug 1321544 - Set a different icon for a contextMenus item](https://bugzilla.mozilla.org/show_bug.cgi?id=1321544)</del>
- <del>[Bug 1325757 - Ability to set label on contextMenu submenu](https://bugzilla.mozilla.org/show_bug.cgi?id=1325757)</del>

WONTFIX'd by Mozilla:
- <del>[Bug 1325758 - Ability to specify insertion point for contextMenu item or submenu](https://bugzilla.mozilla.org/show_bug.cgi?id=1325758)</del>

**Comment by sojusnik** [comment]
If it's possible, you should also consider rearranging the search engines from a list view to a square view, like in [this](https://addons.mozilla.org/de/firefox/addon/quickcontextsearch/) add-on.

It's much easier to handle a lot of search engines that way.

**Comment by ccprog** [comment]
It seems to me that a resolution to [Bug 1352598 - Add an API to list installed search engines](https://bugzilla.mozilla.org/show_bug.cgi?id=1352598) (a pure read-only API) would unblock [Bug 1268401](https://bugzilla.mozilla.org/show_bug.cgi?id=1268401).

**Comment by Taurean75** [comment]
So, is this going to be converted into an WebExtension?

Thanks.

**Comment by gcp** [comment]
It can't be converted as long as bug 1268401 or bug 1352598 aren't fixed, AFAIK.

**Comment by benbasson** [comment]
1268401 isn't really necessary, but 1352598 is 100% required or this add-on isn't going to be able to do much apart from add an empty submenu.

**Comment by benbasson** [comment]
I've updated the issue to reflect the bugs that have been resolved.

**Comment by benbasson** [comment]
Well neither [Bug 1268401](https://bugzilla.mozilla.org/show_bug.cgi?id=1268401) nor [Bug 1352598](https://bugzilla.mozilla.org/show_bug.cgi?id=1352598) is going to be resolved for Firefox 57, so I guess this port isn't happening any time soon. Clearly 2 years wasn't enough to get that API added to WebExtensions.

I'm not doing any more work on this until I get an email regarding one or both of those bugs being resolved. Until then, feel free to lobby Mozilla for that because I'm done with them.

**Comment by mathieumd** [comment]
As an alternative, the [Context Search WebExtension](https://addons.mozilla.org/firefox/addon/context-search-we/) extension uses bookmarks presents into a specific bookmark folder. It's a nice idea, because these are synced between devices through Mozilla Sync.

**Comment by benbasson** [comment]
Well I _could_ do that, but seeing as someone else has already done it I can't really see the point.
